When a macro isn't found, suggest calling a function with the same name

Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7a15474e4972ff683b6d6c7208a5b37f)):

```rust
fn main() {
    std::process::abort!();
}
```

The current output is:

```
error[E0433]: failed to resolve: could not find `abort` in `process`
 --> src/main.rs:2:19
  |
2 |     std::process::abort!();
  |                   ^^^^^ could not find `abort` in `process`

For more information about this error, try `rustc --explain E0433`.
```

Ideally the output should include:

```
std::process::abort is not a macro

help: remove `!` to call the function
```

### Motivation

Note that the opposite mistake, trying to use a macro without the `!`:

```rust
fn main() {
    println("Hello")
}
```

gives:

```
error[E0423]: expected function, found macro `println`
 --> src/main.rs:2:5
  |
2 |     println("Hello")
  |     ^^^^^^^ not a function
  |
help: use `!` to invoke the macro
  |
2 |     println!("Hello")
  |            +

For more information about this error, try `rustc --explain E0423`.
```

### Related issues

This would complement #102601 (suggesting alternative paths when a macro isn't found)

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"chenyukang"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_resolve/src/macros.rs
+++ b/compiler/rustc_resolve/src/macros.rs
@@ -12,7 +12,7 @@ use rustc_attr::StabilityLevel;
 use rustc_data_structures::fx::FxHashSet;
 use rustc_data_structures::intern::Interned;
 use rustc_data_structures::sync::Lrc;
-use rustc_errors::struct_span_err;
+use rustc_errors::{struct_span_err, Applicability};
 use rustc_expand::base::{Annotatable, DeriveResolutions, Indeterminate, ResolverExpand};
 use rustc_expand::base::{SyntaxExtension, SyntaxExtensionKind};
 use rustc_expand::compile_declarative_macro;
@@ -694,7 +694,25 @@ impl<'a> Resolver<'a> {
                     check_consistency(self, &path, path_span, kind, initial_res, res)
                 }
                 path_res @ PathResult::NonModule(..) | path_res @ PathResult::Failed { .. } => {
+                    let mut suggestion = None;
                     let (span, label) = if let PathResult::Failed { span, label, .. } = path_res {
+                        // try to suggest if it's not a macro, maybe a function
+                        if let PathResult::NonModule(partial_res) =  self.resolve_path(
+                            &path,
+                            Some(ValueNS),
+                            &parent_scope,
+                            Some(Finalize::new(ast::CRATE_NODE_ID, path_span)),
+                            None,
+                        ) && partial_res.unresolved_segments() == 0 {
+                            let sm = self.session.source_map();
+                            let span = sm.span_extend_while(span, |c| c == '!').unwrap_or(span);
+                            let code = sm.span_to_snippet(span).unwrap();
+                            suggestion = Some(
+                                    (vec![(span, code.trim_end_matches('!').to_string())],
+                                    format!("{} is not a macro, but a {}, try to remove `!`", Segment::names_to_string(&path), partial_res.base_res().descr()),
+                                    Applicability::MaybeIncorrect)
+                                );
+                        }
                         (span, label)
                     } else {
                         (
@@ -708,7 +726,7 @@ impl<'a> Resolver<'a> {
                     };
                     self.report_error(
                         span,
-                        ResolutionError::FailedToResolve { label, suggestion: None },
+                        ResolutionError::FailedToResolve { label, suggestion },
                     );
                 }
                 PathResult::Module(..) | PathResult::Indeterminate => unreachable!(),

--- a/compiler/rustc_resolve/src/macros.rs
+++ b/compiler/rustc_resolve/src/macros.rs
@@ -697,21 +697,15 @@ impl<'a> Resolver<'a> {
                     let mut suggestion = None;
                     let (span, label) = if let PathResult::Failed { span, label, .. } = path_res {
                         // try to suggest if it's not a macro, maybe a function
-                        if let PathResult::NonModule(partial_res) =  self.resolve_path(
-                            &path,
-                            Some(ValueNS),
-                            &parent_scope,
-                            Some(Finalize::new(ast::CRATE_NODE_ID, path_span)),
-                            None,
-                        ) && partial_res.unresolved_segments() == 0 {
+                        if let PathResult::NonModule(partial_res) = self.maybe_resolve_path(&path, Some(ValueNS), &parent_scope)
+                            && partial_res.unresolved_segments() == 0 {
                             let sm = self.session.source_map();
-                            let span = sm.span_extend_while(span, |c| c == '!').unwrap_or(span);
-                            let code = sm.span_to_snippet(span).unwrap();
-                            suggestion = Some(
-                                    (vec![(span, code.trim_end_matches('!').to_string())],
+                            let exclamation_span = sm.next_point(span);
+                            suggestion = Some((
+                                vec![(exclamation_span, "".to_string())],
                                     format!("{} is not a macro, but a {}, try to remove `!`", Segment::names_to_string(&path), partial_res.base_res().descr()),
-                                    Applicability::MaybeIncorrect)
-                                );
+                                    Applicability::MaybeIncorrect
+                                ));
                         }
                         (span, label)
                     } else {

--- a/src/test/ui/suggestions/issue-103112.rs
+++ b/src/test/ui/suggestions/issue-103112.rs
@@ -1,0 +1,4 @@
+fn main() {
+    std::process::abort!();
+    //~^ ERROR: failed to resolve
+}

--- a/src/test/ui/suggestions/issue-103112.stderr
+++ b/src/test/ui/suggestions/issue-103112.stderr
@@ -1,0 +1,14 @@
+error[E0433]: failed to resolve: could not find `abort` in `process`
+  --> $DIR/issue-103112.rs:2:19
+   |
+LL |     std::process::abort!();
+   |                   ^^^^^ could not find `abort` in `process`
+   |
+help: std::process::abort is not a macro, but a function, try to remove `!`
+   |
+LL |     std::process::abort();
+   |                   ~~~~~
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0433`.

--- a/src/test/ui/suggestions/issue-103112.stderr
+++ b/src/test/ui/suggestions/issue-103112.stderr
@@ -6,8 +6,9 @@ LL |     std::process::abort!();
    |
 help: std::process::abort is not a macro, but a function, try to remove `!`
    |
-LL |     std::process::abort();
-   |                   ~~~~~
+LL -     std::process::abort!();
+LL +     std::process::abort();
+   |
 
 error: aborting due to previous error
 

